src/librustc_trans/save/dump_visitor.rs

```diff
@@ -0,0 +1,1322 @@
+// Copyright 2015 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use super::{escape, generated_code, PathCollector, SaveContext};
+
+use session::Session;
+
+use middle::def::Def;
+use middle::def_id::DefId;
+use middle::ty;
+
+use std::collections::HashSet;
+
+use syntax::ast::{self, CrateNum, NodeId, PatKind};
+use syntax::codemap::*;
+use syntax::parse::token::{self, keywords};
+use syntax::visit::{self, Visitor};
+use syntax::print::pprust::{path_to_string, ty_to_string};
+use syntax::ptr::P;
+
+use rustc_front::lowering::{lower_expr, LoweringContext};
+
+use super::data::*;
+use super::span_utils::SpanUtils;
+
+macro_rules! down_cast_data {
+    ($id:ident, $kind:ident, $this:ident, $sp:expr) => {
+        let $id = if let super::Data::$kind(data) = $id {
+            data
+        } else {
+            $this.sess.span_bug($sp, &format!("unexpected data kind: {:?}", $id));
+        }
+    };
+}
+
+pub trait Dump {
+    // FIXME: create a CrateData struct?
+    fn crate_str(&mut self, span: Span, name: &str, crate_root: &str);
+    // FIXME: create a ExternalCrateData struct?
+    fn external_crate_str(&mut self, name: &str, number: CrateNum, lo_loc: &Loc);
+    // FIXME: this seems like a hack to appease a 3rd party application...
```

Comment:
Indeed. It might be better to have a 'prelude' function (although that name is a bit overloaded already) which would emit the crate_str, external crates, etc. The format for CSV could include the end_external_crates string, whereas JSON could have an array of external_crate_strs instead.

There should be a CratePreludeData struct or something for this.